### PR TITLE
Reimplement PSA and LSA calculations

### PR DESCRIPTION
**Pull request by cjayb**
CC: @ntolley @jasmainak @rythorpe I need your input on this before merging into the main development branch of PR329

I've given the extracellular calculations some thought, and want to propose a slight departure from the original implementation. Note that we began with a flawed one: the transmembrane currents have to be collated at all segments, not just the mid-point (0.5). I fixed that by accounting for all currents, but still kept the "transfer resistances" equal for all segments in a single section.

What I propose here is that transfer resistances (essentially: weighted distances) for both the PSA and LSA be calculated for each segment (see doodle below for LSA). It requires a little more calculation during `__init__`, but adds nothing to the runtime (because the number of segments to keep track of at each time step remains the same). WDYT?

I believe this is what Parasuram et al. (2016) describe in their Figure 1.

I think this is important because our sections are very long, so electrodes placed into the network will be highly dependent on their position wrt. `section(0.5)`, which is unrealistic.

Screenshot 2021-06-01 at 12 26 40

### CODE REVIEW

**Comment by jasmainak** [comment]
I agree this looks more correct. Any before and after plots to share?

**Comment by jasmainak** [comment]
I agree this looks more correct. Any before and after plots to share?

**Comment by cjayb** [comment]
> Any before and after plots to share?

I'll make some...

File: hnn_core/extracellular.py [hunk 1]
```diff
@@ -65,23 +65,30 @@ def _transfer_resistance(section, electrode_pos, sigma, method):
...
+    seg_ctr = np.zeros((section.nseg, 3))
 
-    start = np.array([section.x3d(0), section.y3d(0), section.z3d(0)])
-    end = np.array([section.x3d(1), section.y3d(1), section.z3d(1)])
+    sec_start = np.array([section.x3d(0), section.y3d(0), section.z3d(0)])
+    sec_end = np.array([section.x3d(1), section.y3d(1), section.z3d(1)])
+    sec_ori = sec_end - sec_start
+
+    for ii, seg in enumerate(section):
+        seg_ctr[ii, :] = sec_start + seg.x * sec_ori
```
Comment (jasmainak) [inline hnn_core/extracellular.py:79]: i'm a bit confused by this formulation. Is `sec_ori` a unit vector?

Comment (jasmainak) [reply hnn_core/extracellular.py:79]: oh `seg.x` is between 0 and 1 ...

Comment (cjayb) [reply hnn_core/extracellular.py:79]: Busted. It was originally, but L79 would then read `... = sec_start + seg.x * sec_ori * norm(sec_end - sec_start)` so I just removed the `norm/norm`...

**Comment by cjayb** [comment]
In the below, I plot PSA (blue) and LSA (orange) for the 12 electrodes in the WIP-example ('shank1').

The new implementation leads to mostly identical traces, whereas more PSA vs. LSA differences are present in PR329. This is to be expected from the coarse spatial sampling of membrane currents in the latter.

### PR329
psa_lsa_comp_pr329

### This reimplementation-PR
psa_lsa_comp_reimplementation

File: hnn_core/extracellular.py [hunk 2]
```diff
@@ -65,23 +65,30 @@ def _transfer_resistance(section, electrode_pos, sigma, method):
...
+        # distance from segment midpoint to electrode
+        dis = norm(np.tile(electrode_pos, (section.nseg, 1)) - seg_ctr,
+                   axis=1)
 
-        # setting radius limit
-        if dis < section.diam / 2.0:
-            dis = section.diam / 2.0 + 0.1
+        # # setting radius limit
+        # if dis < section.diam / 2.0:
+        #     dis = section.diam / 2.0 + 0.1
```
Comment (cjayb) [inline hnn_core/extracellular.py:91]: This needs to be discussed. It's related to the discussion/Issue related to cell morphology and spacing.

File: hnn_core/extracellular.py (hunk 1 above)
Comment (jasmainak) [reply hnn_core/extracellular.py:79]: I actually like this formulation better. Also avoids the division by `sec_ori` down below. Anyhow, either works :)

**Comment by jasmainak** [comment]
Just a crazy thought @cjayb. What do you think of dropping support for LSA altogether now that we do the calculations segment-wise? Less maintenance burden ... I don't think any other library supports it anyway.

**Comment by cjayb** [comment]
I'm inclined to agree. There may be a difference at close range, but that "closeness" is the more interesting discussion anyway, related to the spatial scale of our network.

**Pull request merged**